# Imports:

In [164]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Loading Dataset:

In [165]:
credit_risk_dataset = pd.read_csv("../data/raw/credit_risk.csv", index_col=0)

# First look on data:

In [166]:
credit_risk_dataset.head(3)

,SeriousDlqin2yrs,RevolvingUtilizationOfUnsecuredLines,age,NumberOfTime30-59DaysPastDueNotWorse,DebtRatio,MonthlyIncome,NumberOfOpenCreditLinesAndLoans,NumberOfTimes90DaysLate,NumberRealEstateLoansOrLines,NumberOfTime60-89DaysPastDueNotWorse,NumberOfDependents
1,1,0.766127,45,2,0.802982,9120.0,13,0,6,0,2.0
2,0,0.957151,40,0,0.121876,2600.0,4,0,0,0,1.0
3,0,0.658180,38,1,0.085113,3042.0,2,1,0,0,0.0


In [167]:
credit_risk_dataset.describe()

,SeriousDlqin2yrs,RevolvingUtilizationOfUnsecuredLines,age,NumberOfTime30-59DaysPastDueNotWorse,DebtRatio,MonthlyIncome,NumberOfOpenCreditLinesAndLoans,NumberOfTimes90DaysLate,NumberRealEstateLoansOrLines,NumberOfTime60-89DaysPastDueNotWorse,NumberOfDependents
count,150000.000000,150000.000000,150000.000000,150000.000000,150000.000000,1.202690e+05,150000.000000,150000.000000,150000.000000,150000.000000,146076.000000
mean,0.066840,6.048438,52.295207,0.421033,353.005076,6.670221e+03,8.452760,0.265973,1.018240,0.240387,0.757222
std,0.249746,249.755371,14.771866,4.192781,2037.818523,1.438467e+04,5.145951,4.169304,1.129771,4.155179,1.115086
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.000000,0.029867,41.000000,0.000000,0.175074,3.400000e+03,5.000000,0.000000,0.000000,0.000000,0.000000
50%,0.000000,0.154181,52.000000,0.000000,0.366508,5.400000e+03,8.000000,0.000000,1.000000,0.000000,0.000000
75%,0.000000,0.559046,63.000000,0.000000,0.868254,8.249000e+03,11.000000,0.000000,2.000000,0.000000,1.000000
max,1.000000,50708.000000,109.000000,98.000000,329664.000000,3.008750e+06,58.000000,98.000000,54.000000,98.000000,20.000000


In [168]:
credit_risk_dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 150000 entries, 1 to 150000
Data columns (total 11 columns):
 #   Column                                Non-Null Count   Dtype  
---  ------                                --------------   -----  
 0   SeriousDlqin2yrs                      150000 non-null  int64  
 1   RevolvingUtilizationOfUnsecuredLines  150000 non-null  float64
 2   age                                   150000 non-null  int64  
 3   NumberOfTime30-59DaysPastDueNotWorse  150000 non-null  int64  
 4   DebtRatio                             150000 non-null  float64
 5   MonthlyIncome                         120269 non-null  float64
 6   NumberOfOpenCreditLinesAndLoans       150000 non-null  int64  
 7   NumberOfTimes90DaysLate               150000 non-null  int64  
 8   NumberRealEstateLoansOrLines          150000 non-null  int64  
 9   NumberOfTime60-89DaysPastDueNotWorse  150000 non-null  int64  
 10  NumberOfDependents                    146076 non-null  float64
dtypes: fl

In [169]:
credit_risk_dataset.shape

(150000, 11)

## Initial Data Overview

**Shape:** 150,000 rows, 11 columns

**Numerical (int) columns (7):**
1. `delinquent_2yrs` — target variable
2. `age`
3. `late_30_59_count`
4. `open_credit_lines`
5. `late_90_count`
6. `real_estate_loans`
7. `late_60_89_count`

**Numerical (float) columns (4):**
1. `credit_utilization`
2. `debt_ratio`
3. `monthly_income`
4. `dependents`

**Missing values:**
- `monthly_income`: 29,731 missing (~19.8%)
- `dependents`: 3,924 missing (~2.6%)

**Duplicated rows:** 609

Several columns also show suspicious extreme values (`age` min = 0, `debt_ratio` and `credit_utilization` extreme maximums, sentinel-like values in the "late payment" columns). These are investigated in the Data Cleaning section below.



# Rename columns:

In [170]:
credit_risk_dataset.columns = ['delinquent_2yrs', 'credit_utilization', 'age', 'late_30_59_count', 'debt_ratio', 'monthly_income', 'open_credit_lines', 'late_90_count', 'real_estate_loans', 'late_60_89_count', 'dependents']
credit_risk_dataset.head(3)

,delinquent_2yrs,credit_utilization,age,late_30_59_count,debt_ratio,monthly_income,open_credit_lines,late_90_count,real_estate_loans,late_60_89_count,dependents
1,1,0.766127,45,2,0.802982,9120.0,13,0,6,0,2.0
2,0,0.957151,40,0,0.121876,2600.0,4,0,0,0,1.0
3,0,0.658180,38,1,0.085113,3042.0,2,1,0,0,0.0


In [171]:
credit_risk_dataset.select_dtypes(include=["int64"]).columns

Index(['delinquent_2yrs', 'age', 'late_30_59_count', 'open_credit_lines',
       'late_90_count', 'real_estate_loans', 'late_60_89_count'],
      dtype='object')

In [172]:
credit_risk_dataset.select_dtypes(include=["float64"]).columns

Index(['credit_utilization', 'debt_ratio', 'monthly_income', 'dependents'], dtype='object')

In [173]:
credit_risk_dataset.isnull().sum()

delinquent_2yrs           0
credit_utilization        0
age                       0
late_30_59_count          0
debt_ratio                0
monthly_income        29731
open_credit_lines         0
late_90_count             0
real_estate_loans         0
late_60_89_count          0
dependents             3924
dtype: int64

## Data Cleaning

Before diving into individual feature anomalies, we start by removing exact duplicate rows, then investigate suspicious values found during the initial overview.

### Duplicate Rows

In [174]:
credit_risk_dataset.duplicated().sum()

np.int64(609)

In [175]:
credit_risk_dataset[credit_risk_dataset.duplicated()]

,delinquent_2yrs,credit_utilization,age,late_30_59_count,debt_ratio,monthly_income,open_credit_lines,late_90_count,real_estate_loans,late_60_89_count,dependents
1670,0,1.0,29,0,0.0,NaN,0,0,0,0,0.0
7824,0,1.0,29,0,0.0,NaN,0,0,0,0,0.0
7921,0,1.0,22,0,0.0,820.0,1,0,0,0,0.0
8841,0,1.0,23,0,0.0,820.0,1,0,0,0,0.0
10870,0,1.0,73,0,0.0,NaN,2,0,0,0,0.0
...,...,...,...,...,...,...,...,...,...,...,...
149140,0,0.0,77,0,0.0,NaN,2,0,0,0,0.0
149432,0,1.0,25,0,0.0,NaN,1,0,0,0,NaN
149473,0,1.0,84,0,0.0,NaN,1,0,0,0,0.0
149770,0,1.0,23,98,0.0,NaN,0,98,0,98,NaN


In [176]:
credit_risk_dataset = credit_risk_dataset.drop_duplicates()

### Age Anomaly

The initial overview showed a minimum age of 0, which is not a valid value for someone holding a credit line. We check how many rows are affected before deciding how to handle them.

In [177]:
(credit_risk_dataset["age"] == 0).sum()

np.int64(1)

In [178]:
credit_risk_dataset = credit_risk_dataset[credit_risk_dataset["age"] != 0]

### Late Payment Columns — Sentinel Values

The three "late payment" columns (`late_30_59_count`, `late_60_89_count`, `late_90_count`) all share the same maximum value of 98, which is suspicious for a count variable. We check whether values of 96 and 98 appear consistently across all three columns for the same rows, which would suggest these are placeholder/error codes rather than genuine payment counts.

In [179]:
(credit_risk_dataset["late_30_59_count"].isin([96,98])).sum()

np.int64(225)

In [180]:
(credit_risk_dataset["late_90_count"].isin([96,98])).sum()

np.int64(225)

In [181]:
(credit_risk_dataset["late_60_89_count"].isin([96,98])).sum()

np.int64(225)

In [182]:
(credit_risk_dataset["late_30_59_count"].isin([96,98]) & (credit_risk_dataset["late_90_count"].isin([96,98])) & credit_risk_dataset["late_60_89_count"].isin([96,98])).sum()

np.int64(225)

In [183]:
credit_risk_dataset["has_late_data_anomaly"] = credit_risk_dataset["late_30_59_count"].isin([96,98]) & credit_risk_dataset["late_90_count"].isin([96,98]) & credit_risk_dataset["late_60_89_count"].isin([96,98])

### Debt Ratio — Extreme Values

`debt_ratio` shows an extremely skewed distribution (median ≈ 0.37, max ≈ 329,664). Rather than blindly applying an IQR-based cutoff, we first investigate *why* these extreme values occur, since `debt_ratio` is mathematically defined as debt divided by income — a near-zero or missing income can inflate this ratio artificially.

In [184]:
Q1 = credit_risk_dataset['debt_ratio'].quantile(0.25)
Q3 = credit_risk_dataset['debt_ratio'].quantile(0.75)
IQR = Q3 - Q1
upper_bound = Q3 + 1.5 * IQR
print(f"Q1 = {Q1}, Q3 = {Q3}, IQR = {IQR}, upper_bound = {upper_bound}")

Q1 = 0.177441007, Q3 = 0.8752940652500001, IQR = 0.6978530582500001, upper_bound = 1.9220736526250002


In [185]:
credit_risk_dataset.loc[credit_risk_dataset.debt_ratio > upper_bound, 'monthly_income']

7         NaN
9         NaN
15        0.0
17        NaN
33        NaN
         ... 
149977    NaN
149978    NaN
149985    NaN
149993    NaN
149998    NaN
Name: monthly_income, Length: 31285, dtype: float64

### Missing Values

`monthly_income` and `dependents` contain missing values. For each, we first create a binary flag column marking where the original value was missing (preserving that information as a potential signal), then fill the gap using the median — chosen over the mean because both distributions are right-skewed and the median is more robust to extreme outliers.

In [186]:
credit_risk_dataset["monthly_income_nan"] = credit_risk_dataset["monthly_income"].isna()

In [187]:
credit_risk_dataset['monthly_income'] = credit_risk_dataset['monthly_income'].fillna(credit_risk_dataset['monthly_income'].median())

In [188]:
credit_risk_dataset['dependents_nan'] = credit_risk_dataset['dependents'].isna()

In [189]:
credit_risk_dataset['dependents'] = credit_risk_dataset['dependents'].fillna(credit_risk_dataset['dependents'].median())

## Data Cleaning Summary

- Removed 609 exact duplicate rows
- Removed 1 row with an invalid `age` of 0
- Identified and flagged 225 rows with sentinel values (96/98) across all three late-payment columns (`has_late_data_anomaly`)
- Filled missing `monthly_income` (~19.8%) and `dependents` (~2.6%) with the median, preserving missingness via `monthly_income_nan` / `dependents_nan` flags
- Left `debt_ratio` unmodified after confirming its extreme values are largely explained by missing/near-zero income rather than being an independent data issue

**Final shape:** 149,390 rows × 14 columns

In [190]:
credit_risk_dataset.shape

(149390, 14)